### PR TITLE
Add provider pattern

### PR DESCRIPTION
**Pull request by ayeniblessing101**
Hi Dima

Apologies for the delay. I wanted to be sure I understood what I was implementing before I go ahead.

So I was able to come up with something. I created a file for the provider then called it in my handlers but I keep getting an error.

image

### CODE REVIEW

**Comment by dkotik** [comment]
I could not reproduce the error initially. When you send an error, please also send the steps on how you got to it. I fixed up and pushes a commit for you to examine - the source of the problem was that the provider was initialized -before- the database.

`var` statements go before all function calls, even before `init` calls, so if you are initializing things, do it in roughly the same place, so you can enforce the order. I also did a function-with-in-a-function wrap - a really neat technique that is good to learn for passing drivers.

If you are passing many things, a struct is always better.

**Comment by ayeniblessing101** [comment]
Hi @dkotik so I was able to implement a provider to delete a category. I wanted to work on the provider to update category but I curious as to why the method is taking in a struct instead of an ID

**Comment by dkotik** [comment]
> Hi @dkotik so I was able to implement a provider to delete a category. I wanted to work on the provider to update category but I curious as to why the method is taking in a struct instead of an ID

@ayeniblessing101 - it is really up to you how to set it up. An id would work. I like making my APIs as explicit as possible and as similar to each other as possible. So if update method takes a struct, it is easier for me to remember to use a struct for delete as well. The less human memory work - the better.

Much of this is your personal convenience - write for yourself first, the way it is easier for you to remember without having to look up the interface. With experience, you will develop some preferences.

**Comment by ayeniblessing101** [comment]
@dkotik so I made an update to the CategoryUpdate. 

I am used to using ID to update. I took in the category of type *model.Category which will be gotten through the bodyParser which will bind the request body to the category struct

I have a question of as to how to update multiple fields in the stmt.Exec method

File: internal/handlers/categoryHandler.go [hunk 1]
```diff
@@ -53,6 +56,60 @@ func GetCategories(c *fiber.Ctx) error {
...
+		result = append(result, category)
+		return c.Render("category", result)
+	}
+}
+
+// UpdateCategory method update a category from the categories table and returns an error
+func UpdateCategory(p providers.CategoryProvider) func(c *fiber.Ctx) error {
+	return func(c *fiber.Ctx) error {
+		id := c.Params("id")
+		categoryID, _ := strconv.Atoi(id)
```
Comment (dkotik) [inline internal/handlers/categoryHandler.go]: this line could be lower: cat.ID, _ := strconv.Atoi(id) - then you wouldn't need a second ID parameter. But the way you wrote it is ok.

**Review by dkotik — APPROVED** [review]
Good work!

**Comment by ayeniblessing101** [comment]
@dkotik thanks mentor

**Pull request merged**